tests/test_samples.py

```diff
@@ -0,0 +1,170 @@
+import json
+import os
+import threading
+import subprocess
+
+from tests.test_base import BaseEpoClientTest
+from tests.test_value_constants import *
+from tests.mock_eposerver import MockEpoServer
+
+SAMPLE_FOLDER = str(os.path.dirname(
+    os.path.dirname(
+        os.path.abspath(__file__)
+    )
+).replace("\\", "/")) + "/sample"
+
+BASIC_FOLDER = SAMPLE_FOLDER + "/basic"
+
+COMMON_PY_FILENAME = SAMPLE_FOLDER + "/common.py"
+NEW_COMMON_PY_FILENAME = COMMON_PY_FILENAME + ".new"
+
+CONFIG_FOLDER = str(os.path.dirname(os.path.abspath(__file__)).replace("\\", "/"))
+CONFIG_FILENAME = CONFIG_FOLDER + "/dxlclient.config"
+NEW_CONFIG_FILENAME = CONFIG_FILENAME + ".new"
+CA_FILENAME = CONFIG_FOLDER + "/ca-bundle.crt"
+CERT_FILENAME = CONFIG_FOLDER + "/client.crt"
+KEY_FILENAME = CONFIG_FOLDER + "/client.key"
+
+def overwrite_file_line(filename, target, replacement):
+    base_file = open(filename, 'r')
+    new_file = open(filename + ".new", "w+")
+
+    for line in base_file:
+        if line.startswith(target):
+            line = replacement
+        new_file.write(line)
+
+    base_file.close()
+    new_file.close()
+
+    os.remove(filename)
+    os.rename(filename + ".new", filename)
+
+
+def overwrite_common_py():
+    target_line = "CONFIG_FILE = "
+    replacement_line = target_line + "\"" + CONFIG_FILENAME + "\"\n"
+    overwrite_file_line(COMMON_PY_FILENAME, target_line, replacement_line)
+
+
+def overwrite_config_cert_locations():
+    target_line = "BrokerCertChain = "
+    replacement_line = target_line + "\"" + CA_FILENAME + "\"\n"
+    overwrite_file_line(CONFIG_FILENAME, target_line, replacement_line)
+
+    target_line = "CertFile = "
+    replacement_line = target_line + "\"" + CERT_FILENAME + "\"\n"
+    overwrite_file_line(CONFIG_FILENAME, target_line, replacement_line)
+
+    target_line = "PrivateKey = "
+    replacement_line = target_line + "\"" + KEY_FILENAME + "\"\n"
+    overwrite_file_line(CONFIG_FILENAME, target_line, replacement_line)
+
+class SampleRunner(object):
+
+    def __init__(self, cmd, target=''):
+        self.cmd = cmd
+        self.target_file = target
+        self.process = None
+        self.output = "Not started"
+
+
+    def run(self, timeout):
+        def target():
+            self.process = subprocess.Popen(
+                [self.cmd, self.target_file],
+                stdout=subprocess.PIPE,
+                #stderr=subprocess.PIPE,
+            )
+            self.output = self.process.communicate()[0]
+
+        thread = threading.Thread(target=target)
+        thread.start()
+
+        thread.join(timeout)
+        if thread.is_alive():
+            self.process.terminate()
+            thread.join()
+
+        return self.output.decode('utf-8')
+
+class TestCoreHelpSample(BaseEpoClientTest):
+
+    def test_corehelp_example(self):
+        # Modify common/config files to work with local ".\test" directory
+        overwrite_common_py()
```

Comment:
One thing that I think might be nice to consider for this is a way we could run the tests without having any version-controlled files modified as a side effect. If the files are modified when running the tests locally, the likelihood of those changes getting inadvertently included in a commit seems to go up.

For the cert related files, it seems like we might be able to avoid changing the common.py file at all and instead let the execution of the sample files just pick up the certs from the sample directory, as they would by default. We could do 2 provision configs - one for the test directory and one for the sample directory - as part of the Travis config for CI.

For the sample file modifications made below, e.g., for modifying the value of constants like `EPO_UNIQUE_ID`, one approach might be to use Python's `NamedTemporaryFile` to copy the current sample source file into a temp file, modify the constants in the temp file, run the sample, and clean up the temp file on exit.

Does this seem reasonable?

Reply:
I'm on board with this, it's actually what I've started doing in the ePO Service unit tests to fix a bug I was encountering there when trying to create the DxlConfig for the sample's DXL clients.